### PR TITLE
Add poppunk visualisations

### PR DESCRIPTION
**Pull request by muppi1993**
`run_poppunk` endpoint now also queues (dependent) jobs for creating microreact & network visualisations. 
Job IDs are stored separately for each job, `get_status` endpoint reports status of all three of them.
So far the `get_result` endpoint only returns cluster assignment result. Sending visualisation results will be in a later PR.

Tests were adapted to changes.

Question:
How can I unit test the functions in visualise.py? I am testing them implicitly in the unit test for `run_poppunk_internal()`, but for testing them directly I was wondering:
- microreact(): this one depends on the results of the previous job, which it gets from redis. How do I test this function isolated from other jobs?
- network(): since this function is only calling an external function, do I need to test it? Especially since it does not have a return value. If I should test, should I test that files were created like I do in the integration tests?

### CODE REVIEW

File: beebop/visualise.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    Ideally we'd want to only display clusters that have a new isolate added.
+    """
+
+    generate_visualisations(
+        query_db=outdir,
+        ref_db=db_paths.db,
+        distances=outdir + '/' + p_hash + '.dists',
+        rank_fit=None,
+        threads=args.visualise.threads,
+        output=outdir + '/network',
```
Comment (richfitz) [inline beebop/visualise.py:86]: Can you please move all these path manipulations through here into a helper class (e.g., our PoppunkFileStore) as it would be good to centralise this.

**Review by richfitz — CHANGES_REQUESTED** [review]
> microreact(): this one depends on the results of the previous job, which it gets from redis. How do I test this function isolated from other jobs?

Approximately the same way that we do with the `run_poppunk` / `run_poppunk_internal` approach, split the microreact function up into bits that do not depend on the queue state:

```
def microreact(p_hash, outdir, db_paths, args):
    current_job = get_current_job(Redis())
    network_result = current_job.dependency.result
    return microreact_internal(network_result, p_hash, outdir, db_paths, args)


def microreact_internal(network_result, p_hash, outdir, db_paths, args):
    queries_clusters = []
    for item in network_result.values():
        queries_clusters.append(item['cluster'])
    ... # and so on
```

You'll need the poppunk/network bit run somewhere of course - you could do that within the test or set up a directory with suitable output to work with once as part of test setup.

Taking it one step further, I suspect that the [`SimpleWorker`](https://python-rq.org/docs/testing/) approach would allow you to test the whole workflow nicely - we do this in some of the equivalent R versions of this.

> network(): since this function is only calling an external function, do I need to test it? Especially since it does not have a return value. If I should test, should I test that files were created like I do in the integration tests.

Yes, you should test that it runs without error and produces the expected file. That way if the upstream package changes (and it has a lot of args so it's not unexpected) we have a good fast unit test to use to identify this and fix it.

File: .gitignore [hunk 2]
```diff
@@ -10,7 +10,11 @@ ignored
 !tests/files/f3d9b387e311d5ab59a8c08eb3545dbb.json
 !*.schema.json
 !args.json
-tests/files/poppunk_output
+tests/files/poppunk_output/*
```
Comment (richfitz) [inline .gitignore:13]: It might be better to put your files that are under version control at a totally different path. )You can also use `git add -f` to bypass the `.gitignore`). but this way leads to mix of generated and "real" files which is rarely what is wanted

File: beebop/filestore.py [hunk 3]
```diff
@@ -41,6 +41,25 @@ def __init__(self, storage_location):
     def output(self, p_hash):
         return str(PurePath(self.output_base, p_hash))
 
+    def output_network(self, p_hash):
```
Comment (richfitz) [inline beebop/filestore.py:44]: thanks, this is much nicer than having it throughout the code

File: tests/generate_visualisation_input.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+storageLocation = './tests/files'
+p_hash = 'unit_test_visualisations'
+fs = PoppunkFileStore(storageLocation)
+outdir = fs.output(p_hash)
+if not os.path.exists(outdir):
+    os.mkdir(outdir)
+db_paths = DatabaseFileStore('./storage/GPS_v4_references')
+with open("./beebop/resources/args.json") as a:
+    args_json = a.read()
+args = json.loads(args_json, object_hook=lambda d: SimpleNamespace(**d))
```
Comment (richfitz) [inline tests/generate_visualisation_input.py:29]: this block of three lines turns up in three places, I suggest you pop it in a function and save yourself some typing :)

File: tests/test_unit.py [hunk 5]
```diff
@@ -71,34 +97,57 @@ def test_run_poppunk_internal():
...
+    assert read_redis("beebop:hash:job:network",
+                      project_hash, redis) == job_ids["network"]
+
+
+def test_get_result_internal(client):
     # queue example job
     redis = Redis()
-    q = Queue(connection=redis)
-    job = q.enqueue(dummy_fct)
+    q = Queue(connection=Redis())
```
Comment (richfitz) [inline tests/test_unit.py:139]: why this change? redis was defined on the previous line?

File: tests/test_unit.py [hunk 6]
```diff
@@ -194,56 +250,3 @@ def test_check_connection():
         app.check_connection(redis_mock)
     redis = Redis()
     app.check_connection(redis)
-
-
-def test_encoder():
```
Comment (richfitz) [inline tests/test_unit.py:199]: why has this been removed?

File: .gitignore (hunk 2 above)
Comment (muppi1993) [reply .gitignore]: I tried to separate this, having (nearly) all generated files in a `results` folder (see this [commit](https://github.com/bacpop/beebop_py/pull/6/commits/a294ce4e694ca4aa63b5fecf53b29e8b3d99336f)).
Main change in the test files is that I'm using 2 different PoppunkFilestores for file input and output. 
The only two test where this was not straightforward are [`test_assign_clusters`](https://github.com/bacpop/beebop_py/blob/a294ce4e694ca4aa63b5fecf53b29e8b3d99336f/tests/test_unit.py#L54) and [`test_microreact_internal`](https://github.com/bacpop/beebop_py/blob/a294ce4e694ca4aa63b5fecf53b29e8b3d99336f/tests/test_unit.py#L73), where input and output cannot be separated easily. 
For `test_assign_clusters` I solved this by copying the files to the output location first to work at this location.
Is this approach sensible and should I do the same for `test_microreact_internal`?

Comment (muppi1993) [reply .gitignore]: Should I do this then for all tests - not just the ones where separation is tricky, to avoid having the two filestores?

File: .github/workflows/pytest.yml [hunk 7]
```diff
@@ -131,6 +146,6 @@ jobs:
     # tmate. However, don't leave them uncommented on merge as that
     # causes failing builds to hang forever.
     #
-    # - name: tmate session
-    #   if: ${{ failure() }}
-    #   uses: mxschmitt/action-tmate@v3
+    - name: tmate session
+      if: ${{ failure() }}
+      uses: mxschmitt/action-tmate@v3
```
Comment (richfitz) [inline .github/workflows/pytest.yml:151]:
```suggestion
    # - name: tmate session
    #   if: ${{ failure() }}
    #   uses: mxschmitt/action-tmate@v3
```

just to make this easy to disable

**Review by richfitz — APPROVED** [review]
